proxy/grpcproxy/lease.go

```diff
@@ -46,42 +95,177 @@ func (lp *leaseProxy) LeaseTimeToLive(ctx context.Context, rr *pb.LeaseTimeToLiv
-	ctx, cancel := context.WithCancel(stream.Context())
-	lc, err := pb.NewLeaseClient(conn).LeaseKeepAlive(ctx)
-	if err != nil {
-		cancel()
-		return err
+func (lp *leaseProxy) LeaseKeepAlive(stream pb.Lease_LeaseKeepAliveServer) (err error) {
+	fmt.Printf("LeaseKeepAlive() new stream %v \n", stream)
+	cctx, ccancel := context.WithCancel(lp.ctx)
+	lps := leaseProxyStream{
+		stream:          stream,
+		client:          lp.client,
+		keepAliveLeases: make(map[int64]chan struct{}),
+		ctx:             cctx,
+		cancel:          ccancel,
 	}
-
+	defer lps.stop()
+	errCh := make(chan error, 1)
+	// leaseProxyStream.run() guarantee tp be closed after LeaseKeepAlive() exits
+	// because the client_to_proxy stream will be closed. hence LeaseKeepAlive()
+	// can't tracks it.
+	// let leaseProxy keep tracks of leaseProxyStream.run() go routine
+	lp.wg.Add(1)
 	go func() {
-		// Cancel the context attached to lc to unblock lc.Recv when
-		// this routine returns on error.
-		defer cancel()
-
-		for {
-			// stream.Recv will be unblock when the loop in the parent routine
-			// returns on error.
-			rr, err := stream.Recv()
-			if err != nil {
-				return
-			}
-			err = lc.Send(rr)
-			if err != nil {
-				return
-			}
-		}
+		defer lp.wg.Done()
+		errCh <- lps.run()
 	}()
 
-	for {
-		rr, err := lc.Recv()
-		if err != nil {
-			return err
+	var leaderc <-chan struct{}
+	if md, ok := metadata.FromContext(stream.Context()); ok {
+		v := md[rpctypes.MetadataRequireLeaderKey]
+		if len(v) > 0 && v[0] == rpctypes.MetadataHasLeader {
+			leaderc = lp.lostLeaderNotify()
+		}
+	}
+	select {
+	case <-cctx.Done():
+		return cctx.Err()
+	case <-leaderc:
+		fmt.Printf("leaseProxy leader lost err(%v) \n", rpctypes.ErrNoLeader)
+		return rpctypes.ErrNoLeader
+	case err = <-errCh:
+		return err
+	}
+}
+
+type leaseProxyStream struct {
+	stream pb.Lease_LeaseKeepAliveServer
+	client *clientv3.Client
+
+	wg sync.WaitGroup
+	// mu protects keepAliveLeases
+	mu sync.RWMutex
+	// keepAliveLeases map keeps track of on-going lease renewal IDs and corresponding chans.
+	// the chan tracks number of keepalive requests received for a given lease ID.
+	// it is used to control number of keepalive responses to be sent back to the client
+	keepAliveLeases map[int64]chan struct{}
+
+	ctx    context.Context
+	cancel func()
+}
+
+func (lps *leaseProxyStream) run() error {
+	defer func() {
+		fmt.Printf("leaseProxyStream %v closed \n", lps.stream)
+	}()
+	for lps.ctx.Err() == nil {
+		fmt.Printf("leaseProxyStream %v receiving msg...\n", lps.stream)
+		rr, err := lps.stream.Recv()
+		fmt.Printf("leaseProxyStream %v received msg %v \n", lps.stream, rr)
+		if err == io.EOF {
+			return nil
 		}
-		err = stream.Send(rr)
 		if err != nil {
 			return err
 		}
+		lps.serveKeepAliveRequest(rr)
+	}
+	return lps.ctx.Err()
+}
+
+func (lps *leaseProxyStream) stop() {
+	fmt.Printf("LeaseKeepAlive() stream %v closing... \n", lps.stream)
+	lps.cancel()
+	lps.wg.Wait()
```

Comment:
if leaderc is closed, what ensures the recv loop is stopped? The stream ctx won't be canceled by lps.cancel()